R/known-sources.R

```diff
@@ -1,9 +1,52 @@
+# This is a reconstructed schema; it may not be completely accurate, but should
+# be serviceably so. For more information on variables, see the sources below
+# available from Fannie Mae. You may need to create an account first here:
+# https://datadynamics.fanniemae.com/data-dynamics/#/reportMenu;category=HP
+#
+# Data dictionary: https://capitalmarkets.fanniemae.com/resources/file/credit-risk/xls/crt-file-layout-and-glossary.xlsx
+# Names from R code here: https://capitalmarkets.fanniemae.com/media/document/zip/FNMA_SF_Loan_Performance_r_Primary.zip
+fannie_mae_schema <- arrow::schema(
+  LOAN_ID = arrow::string(),
+  ACT_PERIOD = arrow::string(),  # date. Monthly reporting period
+  SERVICER = arrow::string(),
+  ORIG_RATE = arrow::float64(),
+  CURRENT_UPB = arrow::float64(),
+  LOAN_AGE = arrow::int32(),
+  REM_MONTHS = arrow::int32(),
+  ADJ_REM_MONTHS = arrow::int32(),
+  MATR_DT = arrow::string(),  # maturity date
+  MSA = arrow::string(),  # Metropolitan Statistical Area code
+  DLQ_STATUS = arrow::string(),  # Int of months, but `X` is a valid value. New versions pad with `0`/`X` to two characters
+  RELOCATION_MORTGAGE_INDICATOR = arrow::string(),
+  Zero_Bal_Code = arrow::string(),  # 0-padded 2 digit ints representing categorical levels, e.g. "01" -> "Prepaid or Matured"
+  ZB_DTE = arrow::string(),  # date
+  LAST_PAID_INSTALLMENT_DATE = arrow::string(),
+  FORECLOSURE_DATE = arrow::string(),
+  DISPOSITION_DATE = arrow::string(),
+  FORECLOSURE_COSTS = arrow::float64(),
+  PROPERTY_PRESERVATION_AND_REPAIR_COSTS = arrow::float64(),
+  ASSET_RECOVERY_COSTS = arrow::float64(),
+  MISCELLANEOUS_HOLDING_EXPENSES_AND_CREDITS = arrow::float64(),
+  ASSOCIATED_TAXES_FOR_HOLDING_PROPERTY = arrow::float64(),
+  NET_SALES_PROCEEDS = arrow::float64(),
+  CREDIT_ENHANCEMENT_PROCEEDS = arrow::float64(),
+  REPURCHASES_MAKE_WHOLE_PROCEEDS = arrow::float64(),
+  OTHER_FORECLOSURE_PROCEEDS = arrow::float64(),
+  NON_INTEREST_BEARING_UPB = arrow::float64(),
+  MI_CANCEL_FLAG = arrow::string(),  # all null
+  RE_PROCS_FLAG = arrow::string(),
+  LOAN_HOLDBACK_INDICATOR = arrow::string(),  # all null
+  SERV_IND = arrow::string()
+)
+
+
 #' Known data files
 #' @export
 known_sources <- list(
   fanniemae_2016Q4 = list(
     url = "https://ursa-qa.s3.amazonaws.com/fanniemae_loanperf/2016Q4.csv.gz",
-    reader = function(file, ...) arrow::read_delim_arrow(file, delim = "|", col_names = FALSE, ...),
+    schema = fannie_mae_schema,
+    reader = function(file, ...) arrow::read_delim_arrow(file, delim = "|", schema = fannie_mae_schema, ...),
```

Comment:
🤔 Do we need to specify `fannie_mae_schema` twice here? Or is it possible to have the schema in `read_delim_arrow()` be populated from the first specification of `schema = fannie_mae_schema`.

It's not a huge deal to have it twice, but it might be nice to reuse it if it isn't too too much work

Reply:
I didn't because it makes the list definition recursive, but because the function body doesn't get evaluated till it gets called, I think it's fine? I'll try it